lib/healthcheck/elasticsearch_index_diskspace_check.rb

```diff
@@ -0,0 +1,60 @@
+require "govuk_app_config"
+
+module Healthcheck
+  # This is a custom check that is called by GovukHealthcheck
+  # See GovukHealthcheck (govuk_app_config/docs/healthchecks.md) for usage info
+  class ElasticsearchIndexDiskspaceCheck
+    def initialize
+      # Tune this to affect the amount of free space we need available as a minimum % before we alert
+      @low_space_threshold = 20
+
+      # For reporting, ignore
+      @percentage = 0
+    end
+
+    def name
+      :elasticsearch_diskspace
+    end
+
+    def status
+      diskspace_free? ? :ok : :critical
+    end
+
+    def message
+      if diskspace_free?
+        "there is enough diskspace for the elasticsearch indices"
+      else
+        "there is not enough diskspace for the elasticsearch indices! Consider running index cleanup rake task"
+      end
+    end
+
+    def details
+      diskspace_free?
+      { "extra" => "total free space remaining on nodes: #{@percentage}%" }
```

Comment:
It looks like the method `diskspace_free?` could be split up. Rather than set instance variables in the initialize method, you could memoize the methods that provide the data for those variables. This would mean your methods don't have side effects, and make it the calls a bit more transparent. For example:

```ruby
LOW_SPACE_THRESHOLD = 20 # percent

def cluster_stats
  @cluster_stats ||= begin
    client = Services.elasticsearch(cluster: Clusters.default_cluster)
    es_stats = client.perform_request "GET", "_nodes/stats/fs?pretty=true"
    es_stats.body.dig("nodes").each_with_object({ total: 0, avail: 0 }) do |(_, node_stat), hsh|
      hsh[:total] += node_stat.dig("fs", "total", "total_in_bytes")
      hsh[:avail] += node_stat.dig("fs", "total", "available_in_bytes")
    end
  end
end

def percentage_free
  cluster_stats[:avail] / (cluster_stats[:total] / 100)
end

def low_diskpace?
  percentage_free < LOW_SPACE_THRESHOLD
end
```

Reply:
Good call, I'm still getting my head around some aspects of Ruby's function handling, took the easy route!